bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/DefaultEntryLogger.java

```diff
@@ -1036,15 +1036,20 @@ public void scanEntryLog(long entryLogId, EntryLogScanner scanner) throws IOExce
                 }
                 // read the entry
                 data.clear();
-                data.capacity(entrySize);
-                int rc = readFromLogChannel(entryLogId, bc, data, pos);
-                if (rc != entrySize) {
-                    LOG.warn("Short read for ledger entry from entryLog {}@{} ({} != {})",
-                            entryLogId, pos, rc, entrySize);
-                    return;
+                int capacity = Math.min(scanner.getLengthToRead(), entrySize);
```

Comment:
if getLengthToRead returns READ_NOTHING then maybe "data" can be passed as "null"
if getLengthToRead returns READ_LEDGER_ENTRY_ID then the caller is interested only in the initial bytes.

Please consider changing the interface and have 3 different "process" methods

READ_NOTHING:
process(ledgerId, entrySize)

READ_LEDGER_ENTRY_ID;
process(ledgerId, entryId, entrySize)
in this case the entryId is extracted by this code and not by the code downstream

READ_ALL
process(ledgerId, offset, data)

This way it is clearer what is happening, and the code is safer, as the implementation of the interface doesn't need to understand the internal format of the "data" buffer when it doesn't need it.
Using the enum (as suggested below) does not introduce extra costs at runtime

Reply:
Good idea, i have refactored the code, please help to review it again, thanks.